File: src/main/java/cn/solarmoon/idyllic_food_diary/feature/basic_feature/IPlateable.java
package cn.solarmoon.idyllic_food_diary.feature.basic_feature;

import cn.solarmoon.idyllic_food_diary.IdyllicFoodDiary;
import cn.solarmoon.idyllic_food_diary.util.ContainerHelper;
import cn.solarmoon.solarmoon_core.api.util.LevelSummonUtil;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.level.block.entity.BlockEntity;

/**
 * 可预存result和container的接口，适用于一些需要将合成结果预存在方块中再用容器取出的配方<br/>
 * 也是容器信息保存的桥梁<br/>
 * 在使用上，在类中给予空的result和container，在配方中找位置调用tryGiveResult/setContainer/setResult
 */
public interface IPlateable extends IExpGiver {

    String CONTAINER = "Container";
    String RESULT = "Result";

    /**
     * 将预存的结果给予玩家<br/>
     * 并且会将玩家手中盛放的容器信息存入食物
     */
    default boolean tryGiveResult(Player player, InteractionHand hand) {
        ItemStack heldItem = player.getItemInHand(hand);
        return tryGiveResult(player, heldItem);
    }

    /**
     * 将预存的结果给予玩家<br/>
     * 并且会将玩家手中盛放的容器信息存入食物
     */
    default boolean tryGiveResult(Player player, ItemStack heldItem) {
        if (hasResult()) {
            if (getContainer().test(heldItem)) {
                ItemStack result = getResult().split(1);
                ContainerHelper.setContainer(result, heldItem); // 这里保存了容器信息
                LevelSummonUtil.addItemToInventory(player, result);
                giveExp(player, true);
                if (!player.isCreative()) heldItem.shrink(1);
                resetContainer();
                ((BlockEntity)this).setChanged();
                return true;
            } else {
                Component message = IdyllicFoodDiary.TRANSLATOR.set("message", "container_required.empty");
                if (!getContainer().isEmpty()) {
                    message = IdyllicFoodDiary.TRANSLATOR.set("message", "container_required");
                }
                player.displayClientMessage(message, true);
            }
        }
        return false;
    }

    default void resetContainer() {
        setContainer(Ingredient.EMPTY);
    }

    /**
     * @return 预留槽是否被占用
     */
    default boolean hasResult() {
        return !getResult().isEmpty();
    }

    /**
     * 同时设置输出结果和匹配容器
     */
    default void setPending(ItemStack result, Ingredient container) {
        setResult(result);
        setContainer(container);
    }

    ItemStack getResult();

    Ingredient getContainer();

    void setResult(ItemStack newResult);

    void setContainer(Ingredient newContainer);

}

File: src/main/kotlin/cn/solarmoon/idyllic_food_diary/element/recipe/assistant/IPlateable.kt
package cn.solarmoon.idyllic_food_diary.element.recipe.assistant

import cn.solarmoon.idyllic_food_diary.IdyllicFoodDiary
import cn.solarmoon.idyllic_food_diary.element.matter.food_container.FoodContainerItem
import cn.solarmoon.idyllic_food_diary.feature.food_container.FoodContainer
import cn.solarmoon.idyllic_food_diary.feature.util.MessageUtil
import cn.solarmoon.idyllic_food_diary.registry.common.IFDDataComponents
import cn.solarmoon.spark_core.api.util.DropUtil
import net.minecraft.core.HolderLookup
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtOps
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.InteractionHand
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.Ingredient
import net.minecraft.world.level.block.entity.BlockEntity

/**
 * 需要预存输出内容的可接入
 */
interface IPlateable: IExpGiver {

    var containerIdentifier: Component
    var result: ItemStack
    var container: Ingredient

    /**
     * 将预存的结果给予玩家
     *
     * 并且会将玩家手中盛放的容器信息存入食物
     */
    fun tryGiveResult(player: Player, hand: InteractionHand, mainHandOnly: Boolean): Boolean {
        if (mainHandOnly && hand != InteractionHand.MAIN_HAND) return false
        val heldItem = player.getItemInHand(hand)
        return tryGiveResult(player, heldItem)
    }

    fun getBlockEntity(): BlockEntity

    /**
     * 将预存的结果给予玩家
     *
     * 并且会将玩家手中盛放的容器信息存入食物
     */
    fun tryGiveResult(player: Player, heldItem: ItemStack): Boolean {
        if (hasResult()) {
            if (container.test(heldItem) && FoodContainerItem.isPureTexture(heldItem, player.level())) {
                val result = result.split(1)
                if (!heldItem.isEmpty) result.set(IFDDataComponents.FOOD_CONTAINER, FoodContainer(heldItem))
                if (!player.isCreative) heldItem.shrink(1)
                DropUtil.addItemToInventory(player, result)
                giveExp(player, true)
                resetContainer()
                getBlockEntity().setChanged()
                return true
            } else {
                val level = player.level()
                if (level.isClientSide) {
                    var message = IdyllicFoodDiary.TRANSLATOR.set("message", "container_required", MessageUtil.EMPTY_HAND)
                    if (!container.isEmpty) {
                        var id: Component
                        if (container.items.size == 1 && containerIdentifier.string.isEmpty()) id = MessageUtil.identifier(container.items[0])
                        else id = containerIdentifier
                        message = IdyllicFoodDiary.TRANSLATOR.set("message", "container_required", id)
                    }
                    player.displayClientMessage(message, true)
                }
            }
        }
        return false
    }

    fun resetContainer() {
        container = Ingredient.EMPTY
    }

    /**
     * @return 预留槽是否被占用
     */
    fun hasResult(): Boolean {
        return !result.isEmpty
    }

    /**
     * 同时设置输出结果和匹配容器
     */
    fun setPending(result: ItemStack, container: Ingredient, id: Component) {
        this.result = result
        this.container = container
        this.containerIdentifier = id
    }

    override fun aSave(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.aSave(tag, registries)
        tag.put("Result", result.saveOptional(registries))
        tag.put("Container", Ingredient.CODEC.encodeStart(registries.createSerializationContext(NbtOps.INSTANCE), container).orThrow)
    }

    override fun aLoad(tag: CompoundTag, registries: HolderLookup.Provider) {
        super.aLoad(tag, registries)
        result = ItemStack.parseOptional(registries, tag.getCompound("Result"))
        container = Ingredient.CODEC.decode(registries.createSerializationContext(NbtOps.INSTANCE), tag.get("Container")).orThrow.first
    }

}